server/src/graphql/mutations/deleteApp.ts

```diff
@@ -0,0 +1,63 @@
+import { MutationResolvers } from '../../generated/graphql';
+import { prisma } from '../../prisma';
+import { dokku } from '../../lib/dokku';
+import { sshConnect } from '../../lib/ssh';
+import { appNameSchema } from '../utils';
+
+export const deleteApp: MutationResolvers['deleteApp'] = async (
+  _,
+  { input },
+  { userId }
+) => {
+  if (!userId) {
+    throw new Error('Unauthorized');
+  }
+
+  // We make sure the name is valid to avoid security risks
+  appNameSchema.validateSync({ name: input.name });
+
+  // TODO check name of the app is unique per server
+
+  const ssh = await sshConnect();
+
+  await dokku.apps.deleteApp(ssh, input.name);
+
+  // We find app to delete
+  const allApps = await prisma.user
+    .findOne({
+      where: {
+        id: userId,
+      },
+    })
+    .App();
+
+  const appToDelete = allApps.filter((app) => app.name === input.name);
+
+  // We find all related app builds and delete them
+  const allAppBuilds = await prisma.user
+    .findOne({
+      where: {
+        id: userId,
+      },
+    })
+    .AppBuild();
+
+  const appBuildToDelete = allAppBuilds.filter(
+    (appBuild) => appBuild.appId === appToDelete[0].id
+  );
+
+  await prisma.appBuild.deleteMany({
+    where: {
+      id: appBuildToDelete[0].id,
+    },
+  });
+
+  // We delete the app itself
+  await prisma.app.delete({
```

Comment:
@Akirtovskis do you know if prisma support cascade delete? If yes that would simplify a lot this function

Reply:
@pradel they currently don't.  It seems that it will be done at some point "soon" which probably means this year as discussed here : https://github.com/prisma/migrate/issues/249 . 

There's a [link](https://prisma2.netlify.app/guides/database-workflows/cascading-deletes/postgresql/#3-create-two-tables-with-a-foreign-key-and-cascade-deletion-behavior) to describe how it can be done, but for me it seems better to wait while they implement this feature. So I added to do for now. 

What's your opinion on this? Better to wait or go with what was suggested in the link?

Reply:
Let's wait and change the logic here once it's ready :+1: